### DISCUSSION TITLE
Preconditioner reuse -- speeding up Newton/LU runs by 10x

### DISCUSSION THREAD

**Discussion by reverendbedford** [idaholab/moose#20627] [Q&A General] [answered]
For better or worse some of us run apps using a full Newton solve and superlu_dist to solve the linear equations.  Typically in my group here at ANL we run "medium sized" jobs like this all the time on workstations.  They might take a few hours to a few days and typically involve up to a couple million dofs.  Anything larger and we move to HPC and use an iterative linear solver preconditioned by hypre.  But running jobs with superlu is much faster, assuming you have enough memory to factor the linearized system of equations.

We do a lot of these medium sized problems.  Here's an example -- some RVE type model of a microstructure with a bunch of inclusions providing some strength and we're pulling on it to simulate a flow curve.  I can provide the input file and mesh for those who are curious, but it uses one of the things we have in [deer](https://github.com/Argonne-National-Laboratory/deer) that's not in tensor_mechanics.

![mesh](https://user-images.githubusercontent.com/7506243/159709208-dcab2075-7ebd-413d-9641-e536a0c9c8b8.png)

If I run this example problem on 24 ranks (~25000 dofs/rank) with the following solver block it takes about 116 minutes to run 10 time steps.

```
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -ksp_type'
  petsc_options_value = 'lu superlu_dist preonly'
```

If I run with this solver block, which involves some new options I added over [here](https://github.com/reverendbedford/moose/tree/precond_reuse) *and needs a hacked version of libmesh*, the exact same problem takes 11 minutes.  So it's 10ish times faster.

```
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -ksp_type'
  petsc_options_value = 'lu superlu_dist gmres'
  l_tol = 1e-8

  reuse_preconditioner = true
  reuse_l_max_its = 50
```

What the new options do is allow `NonlinearProblem` and the associated libmesh `PetscNonlinearSolver` to reuse the LU factored Jacobian from previous nonlinear iterations *and even previous timesteps* as the preconditioner for GMRES to continue to solve the current linearized system of equations.  The heuristic is you just keep reusing your LU decomposition until the number of GMRES iterations exceeds some number -- here 50.  At that point you just refactor the system, do your GMRES (which is just iterative refinement during this first nonlinear iteration), and then keep using that new LU decomposition as a preconditioner for awhile.

This is definitely a heuristic but in my experience it works very, very well because really the structure of the nonlinear equations isn't changing all that much between iterations and changes only very slowly across timesteps.  In this problem with reuse we only factor the linear system only 1 time, versus 30 times for the "base" set.  And I'm not the only one to have noticed how well this works: the Intel PARDISO folks have been using it for 8 years or so and it's supported in PETSc via [SNESSetLagPreconditionerPersists](https://petsc.org/main/docs/manualpages/SNES/SNESSetLagPreconditionerPersists.html) and [SNESSetLagPreconditioner](https://petsc.org/main/docs/manualpages/SNES/SNESSetLagPreconditioner.html), which are of course how I implemented it in the branch.

Reconfiguring things to allow preconditioner reuse within a *single* time step is easy.  Allowing use across *multiple* time steps requires a change to libmesh because right now it [clears](https://github.com/libMesh/libmesh/blob/e10cc1ba8370ccdb761084708d5b6d759894cc12/src/solvers/petsc_nonlinear_solver.C#L1001) the SNES instance in between each time step.  I started a [discussion](https://github.com/libMesh/libmesh/discussions/3201) over in libmesh on changing this behavior, my preliminary tests suggest it's not that hard and doesn't seem to cause any memory issues.

So all that said:

1. Would anyone else find this useful?
2. Does anyone more informed than me foresee problems in not releasing the SNES instance between time steps?  If we go ahead I'll make the old "release after every nonlinear solve" behavior the default, but it would be good to know if MOOSE (or libmesh) somehow relies on this behavior to ensure cross-compatibility with arbitrary apps wanting to use this heuristic.

**Comment by GiudGiud** [comment]
I'm interested. We use Newton/LU for Navier Stokes too currently (we will have a segregated solver soon though). But the problem is pretty non-linear so no guarantees reusing preconditioners would work.

**Comment by hugary1995** [comment]
This would be very useful. Also, defaults could be set on a per-app basis, so that shouldn't create any compatibility issue moving forward -- we could keep the current moose behavior as the default and let apps opt in using this new capability by chaging their default.

**Comment by reverendbedford** [comment, marked as answer]
For those who are interested you can play with [this](https://github.com/reverendbedford/moose/tree/reuse_preconditioner) branch.  Note that it points to a fork of libmesh [here](https://github.com/reverendbedford/libmesh/tree/preconditioner_reuse).  There is a [PR](https://github.com/libMesh/libmesh/pull/3208) open on libmesh to get my changes in over there, then I can make a proper PR on MOOSE.

The solver parameters to play with are

```
  reuse_preconditioner = true
  reuse_preconditioner_max_its = 30
```

and you'll want to use that with something like

```
  petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'gmres lu superlu_dist'
```

**Reply by reverendbedford** [reply]
I should add I'm curious how it will work for other physics outside of tensor_mechanics.

**Comment by bwspenc** [comment]
Agreed that this is useful, and a 10x speedup definitely warrants develping this! Our group also runs in this mode (moderate sized problems with superlu) quite a bit, and I've often wondered how much time we could save by reusing the preconditioner. It seems like this might be even more helpful if you're using PJFNK rather than Newton because in that case, you can often get more mileage with a less-than-perfect preconditioner than you would with Newton.

 I guess it depends on the problem, but I can see the benefit of allowing to reuse the preconditioner between steps. I can't think of anything offhand that would be an issue there, other than the fact that the problem might change and the preconditioner might not be that great. In particular, we often use damage models in which we lag the damage by one step, and I imagine the performance would be hampered considerably by that. I'd just add separate options that control whether to reuse the preconditioner between steps or within the current step only.

**Comment by Traiwit** [comment]
@reverendbedford thanks for sharing!

I find it slower compared to `hypre` for my problem:
Porousflow, highly heterogenous, have many faults (discontinuity), a massive model (6 million elements), complex surface drainage setup.

`hypre` would take 15 mins for 1 steady-step to converge on my workstation, but `reuse_preconditioner` wouldn't converge after 30 mins:
![image](https://user-images.githubusercontent.com/79562439/161458964-f1c3b41b-a4da-4dbe-b720-0d65318f83e3.png)


here is my setup for hypre:

```
[Executioner]
  type = Steady
  solve_type = NEWTON
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre    boomeramg'

  nl_rel_tol = 1e-15
  l_tol = 1e-06
  l_max_its = 150
  nl_max_its = 50
[]
```

maybe I didn't setup `reuse_preconditioner` correctly 

Kind regards,
Traiwit

**Reply by reverendbedford** [reply]
You need to run a transient problem to see if it helps or not (or have a steady problem with a huge number of nonlinear iterations).  It reuses the preconditioner across times steps, which is why I was able to get such good speedup.

**Reply by Traiwit** [reply]
I see, noted, I will give it a go with pit backfill + reflooding problem, and get back to you soon

thank you @reverendbedford

**Reply by reverendbedford** [reply]
Also: I don't know your problems but

```
  nl_rel_tol = 1e-15
```

seems a bit much.  It might have trouble solving the linearized systems in a reasonable amount of iterations when you're trying to converge things down that fine.  But maybe it would be okay as long as you are comparing everything at the same tolerances?

Could you try something like:

```
  petsc_options_iname = '-ksp_type -pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'gmres lu superlu_dist'

  nl_rel_tol = 1e-8
  l_tol = 1e-5
  l_max_its = 50
  nl_max_its = 50

  reuse_preconditioner = true
  reuse_preconditioner_max_its = 30
```

and see if that works for you?

By the way, there's nothing to stop you from using this with hypre.  In your example above you could also just do

```
[Executioner]
  type = Steady
  solve_type = NEWTON
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre    boomeramg'

  nl_rel_tol = 1e-15
  l_tol = 1e-06
  l_max_its = 150
  nl_max_its = 50

  reuse_preconditioner = true
  reuse_preconditioner_max_its = 100
[]
```

and see what happens.
